android/app/src/main/java/com/coboltforge/dontmind/multivnc/VncCanvas.java

```diff
@@ -679,6 +679,29 @@ public boolean processPointerEvent(MotionEvent evt,boolean mouseIsDown,boolean u
 		}
 	}
 
+	/**
+	 * Process given mouse event.
+	 *
+	 * @param button Which button is pressed/released (0 for no button)
+	 * @param isDown True if button is down
+	 * @param x,y    Event position in remote frame buffer coordinates.
+	 */
+	public void processMouseEvent(int button, boolean isDown, int x, int y) {
```

Comment:
what about multiple buttons at once? i.e. merge the `button` and `isDown` args into a bitmask?

Reply:
Each button will generate separate `ACTION_BUTTON_PRESS` & `ACTION_BUTTON_RELEASE` events. `processMouseEvent` will be called multiple time with different `button` value. When `isDown` is true, we set the flag in `pointerMask`. So if two buttons go down `pointerMask` will indicate both of them.

Here is an example of event stream captured on my device (OT = `onTouchEvent`, GM = `onGenericMotionEvent`):

<details>
<summary>Event Log</summary>
<p>

```python
Left click
OT: MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x2, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18650565, downTime=18650565, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_PRESS, actionButton=BUTTON_PRIMARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x2, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18650565, downTime=18650565, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_RELEASE, actionButton=BUTTON_PRIMARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x2, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18650733, downTime=18650565, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_UP, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x2, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18650733, downTime=18650565, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_ENTER, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x2, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18650733, downTime=18650565, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x2, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18650733, downTime=18650565, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_EXIT, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653622, downTime=18653622, deviceId=21, source=0x2002 }


Right click
OT: MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653622, downTime=18653622, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_PRESS, actionButton=BUTTON_SECONDARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653622, downTime=18653622, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_RELEASE, actionButton=BUTTON_SECONDARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653741, downTime=18653622, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_UP, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653741, downTime=18653622, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_ENTER, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653741, downTime=18653622, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18653741, downTime=18653622, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_EXIT, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18656333, downTime=18656333, deviceId=21, source=0x2002 }


Both buttons down
OT: MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18656333, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_PRESS, actionButton=BUTTON_PRIMARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18656333, downTime=18656333, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY|BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18656886, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_PRESS, actionButton=BUTTON_SECONDARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY|BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18656886, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_RELEASE, actionButton=BUTTON_SECONDARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18657222, downTime=18656333, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18657222, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_RELEASE, actionButton=BUTTON_PRIMARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18657669, downTime=18656333, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_UP, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18657669, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_ENTER, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18657669, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18657669, downTime=18656333, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_EXIT, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18659572, downTime=18659572, deviceId=21, source=0x2002 }



Both buttons down, 2nd example
OT: MotionEvent { action=ACTION_DOWN, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18659572, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_PRESS, actionButton=BUTTON_PRIMARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18659572, downTime=18659572, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY|BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18659972, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_PRESS, actionButton=BUTTON_SECONDARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_PRIMARY|BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18659972, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_RELEASE, actionButton=BUTTON_PRIMARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18660053, downTime=18659572, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=BUTTON_SECONDARY, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18660053, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_BUTTON_RELEASE, actionButton=BUTTON_SECONDARY, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18660669, downTime=18659572, deviceId=21, source=0x2002 }
OT: MotionEvent { action=ACTION_UP, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18660669, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_ENTER, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18660669, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_MOVE, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18660669, downTime=18659572, deviceId=21, source=0x2002 }
GM: MotionEvent { action=ACTION_HOVER_EXIT, actionButton=0, id[0]=0, x[0]=395.75833, y[0]=283.4854, toolType[0]=TOOL_TYPE_MOUSE, buttonState=0, metaState=0, flags=0x0, edgeFlags=0x0, pointerCount=1, historySize=0, eventTime=18662187, downTime=18659572, deviceId=21, source=0x2002 }

```

</p>
</details>